### ISSUE TITLE
dev branch: Ledger Nano S issues

### ISSUE THREAD

**Issue by brianlenz** [CityOfZion/neon-wallet#798]
I've done some QA on the `dev` branch and noticed a few quirks and issues with the Ledger as compared to 0.1.4. I don't know if any of these will be addressed by #787, so recording them in case the two aren't related.

**1: Ledger USB Error**

When you attempt to use Ledger without a device connected, it now displays a USB error that wasn't displayed previously:

> Looking for USB Devices. Please plugin your device and login.
> USB Error: No device found.

On 0.1.4, that second line wasn't displayed, which I think is slightly preferable?

**2: Invalid status error**

The message displayed once you are logged has changed:

> Looking for USB Devices. Please plugin your device and login.
> Invalid status 6e00

Previously, on 0.1.4, it was easier to understand the issue:

> Found USB Ledger Nano S
> Hardware Device Error. Login to NEO App and try again

**3: $Ledger $Nano**

Once you have opened the NEO app, you see:

> Found USB $Ledger $Nano S

as opposed to 0.1.4:

> Found USB Ledger Nano S

Really minor, but probably worth removing the dollar signs.

If you had previously gone through issue **2** above, you will still see the "Invalid status 6e00" error at this point, as well.

**4: Can't Sign Some NEP-5 Transfer Transactions**

It looks like signing NEP-5 transfer transactions on the Ledger isn't working 100% reliably. I've seen the blue popup display saying you need to sign the transaction on the device, but then it quickly disappears and the red Transaction Failed message appears. This happened to me 3 times, then I tried transferring NEO instead, which worked. After that, I was able to transfer the NEP-5 token successfully a couple of times, even after restarting Neon. Then, I tried another transfer transaction, which asked me to sign. I signed the transaction on the Ledger and then immediately got the "Transaction Failed" red message again. I can't seem to identify a pattern as to when it succeeds and when it fails, but it's failing about 40% of the time in my tests.

**5: Concurrent Wallets w/ Ledger**

This is probably something we don't care about fixing, but one thing I have been doing in my tests is testing the new build against the current release build in parallel. In the past, this worked without issues, even when using the Ledger Nano S in both. The `dev` branch now reports this error when opening the Ledger Nano S when I try to do it:

> cannot open device with path IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/XHC1@14/XHC1@14000000/HS05@14400000/USB2.0 Hub @14400000/AppleUSB20Hub@14400000/AppleUSB20HubPort@14440000/USB2.0 Hub @14440000/AppleUSB20Hub@14440000/AppleUSB20HubPort@14444000/Nano S@14444000/Nano S@0/IOUSBHostHIDDevice@14444000,0

The 0.1.4 version in this case jumps back and forth between "Looking for USB Devices" and "Found USB Ledger Nano S".

I'm able to successfully open the wallet on both apps concurrently, but they don't show the same information, so it doesn't appear to be supported anymore. I'm willing to concede this isn't a primary use case, but I wanted to mention that it was at least a difference I noticed compared to previous builds. Perhaps at a minimum worth handling the error message more gracefully?

**Comment by mhuggins** [comment]
@brianlenz I believe points 1 & 3 are now resolved in the `dev` branch.

**Comment by brianlenz** [comment]
Confirmed issue **3** is fixed!

Issue **1** still seems to be unchanged (though that's not a huge issue).

New issue (something I probably just didn't notice previously):

**6: Error opening/closing Neo app**

I noticed two errors that are displayed temporarily when loading or unloading the Neo app from the Ledger login screen. When loading the Neo app after first launching Neon:

> Cannot write to HID device

That same message will display if you subsequently quit the Neo app while still on the Ledger login page. I've also seen:

> could not read from HID device

It doesn't happen every time, and it only displays for at most a second or two before disappearing. You should be able to repro it if you just open and quit the Neo app a few times in a row when on the Ledger login screen. It doesn't seem to affect behavior, so perhaps these are transient I/O issues that should just be hidden from display unless the issue remains persistently?

**Comment by mhuggins** [comment]
Point 2 should now be fixed in dev thanks to @snowypowers.

**Comment by brianlenz** [comment]
**Issue 2** looks fixed 👍  I did notice a "cannot open device with path" when first opening the NEO app. Similar to **issue 6**, but a different message.

Also, I saw the same, long error message from **issue 5** when closing the NEO app, so it looks like that message isn't isolated to just the scenario where you have two Neon wallet apps open concurrently.

**Comment by mhuggins** [comment]
@brianlenz I'm curious to see if issue 4 is solved by #804.

Issue 5 is an artifact of switching to their official repo for the ledger NPM package I believe.  I don't think there's anything we can do about it.

**Comment by brianlenz** [comment]
It looks like there is a new issue on the latest `dev` preventing Ledger from working altogether:

**Issue 7: Ledger Login Fails**

Try to login with your Ledger. Upon clicking the "Use Ledger Nano S" button, the UI goes white and there are multiple errors in the console:

```
TRANSACTION_HISTORY request action failed. Error: new BigNumber() not a number: GAS
at raise (http://localhost:3000/dist/bundle.js:32508:25)
    at http://localhost:3000/dist/bundle.js:32496:33
    at new BigNumber (http://localhost:3000/dist/bundle.js:31511:67)
    at toBigNumber (http://localhost:3000/dist/bundle.js:15311:10)
    at http://localhost:3000/dist/bundle.js:2047:177
    at Array.map (native)
    at _callee$ (http://localhost:3000/dist/bundle.js:2039:61)
    at tryCatch (http://localhost:3000/dist/bundle.js:128492:40)
    at Generator.invoke [as _invoke] (http://localhost:3000/dist/bundle.js:128730:22)
    at Generator.prototype.(anonymous function) [as next] (http://localhost:3000/dist/bundle.js:128544:21)
```

```
Uncaught TypeError: Cannot read property 'NEO' of null
    at mapAccountDataToProps (index.js:30)
    at Function.mapStateToProps [as mapToProps] (withData.js:38)
    at mapToPropsProxy (wrapMapToProps.js:43)
    at Function.detectFactoryAndVerify (wrapMapToProps.js:52)
    at mapToPropsProxy (wrapMapToProps.js:43)
    at handleFirstCall (selectorFactory.js:26)
    at pureFinalPropsSelector (selectorFactory.js:74)
    at Object.runComponentSelector [as run] (connectAdvanced.js:26)
    at Connect.initSelector (connectAdvanced.js:178)
    at new Connect (connectAdvanced.js:119)
```

This issue doesn't seem to occur when using a saved wallet, so it appears to be Ledger specific based on my initial tests.

If you need more details from me, let me know!

**Comment by brianlenz** [comment]
I've updated to the Ledger Nano S firmware 1.4.1, and it seems some of these issues are more prevalent now.

**7: Invalid sequence**

This one is new. When the NEO app isn't opened but your Ledger is unlocked, the message displayed jumps back and forth between:

> Your NEO app is closed! Please login.

and

> Invalid sequence

It seems that every 1-3 seconds it flips back and forth between those. It's a bit jarring.

**8: Cannot open device with path**

A bit of a combination between **issue 5** and **issue 7**, when your Ledger is not unlocked, the message displayed jumps back and forth between:

> cannot open device with path IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/XHC1@14/XHC1@14000000/HS05@14400000/USB2.0 Hub @14400000/AppleUSB20Hub@14400000/AppleUSB20HubPort@14440000/USB2.0 Hub @14440000/AppleUSB20Hub@14440000/AppleUSB20HubPort@14444000/Nano S@14444000/Nano S@0/IOUSBHostHIDDevice@14444000,0

and:

> Invalid sequence

Perhaps we should not generically show these messages anymore? Almost every message that I've enumerated in this issue are meaningless to users (and will only cause alarm). Instead, maybe we should detect if the error message is a _known_ message, and only display the custom messaging in that case. In all other cases, it probably makes sense to suppress the message entirely?
